lib/pos_mempool.go

```diff
@@ -0,0 +1,391 @@
+}
+
+// PosMempool is used by the node to keep track of uncommitted transactions. The main responsibilities of the PosMempool
+// include addition/removal of transactions, back up of transaction to database, and retrieval of transactions ordered
+// by Fee-Time algorithm. More on the Fee-Time algorithm can be found in the documentation of TransactionRegister.
+type PosMempool struct {
+	sync.RWMutex
+	status PosMempoolStatus
+	// params of the blockchain
+	params *DeSoParams
+	// globalParams are used to track the latest GlobalParamsEntry. In case the GlobalParamsEntry changes, the PosMempool
+	// is equipped with UpdateGlobalParams method to handle upgrading GlobalParamsEntry.
+	globalParams *GlobalParamsEntry
+	// dir of the directory where the database should be stored.
+	dir string
+	// db is the database that the mempool will use to persist transactions.
+	db *badger.DB
+
+	// txnRegister is the in-memory data structure keeping track of the transactions in the mempool. The TransactionRegister
+	// is responsible for ordering transactions by the Fee-Time algorithm.
+	txnRegister *TransactionRegister
+	// ledger is a simple in-memory data structure that keeps track of cumulative transaction fees in the mempool.
+	// The ledger keeps track of how much each user would have spent in fees across all their transactions in the mempool.
+	ledger *BalanceLedger
+	// persister is responsible for interfacing with the database. The persister backs up mempool transactions so not to
+	// lose them when node reboots. The persister also retrieves transactions from the database when the node starts up.
+	// The persister runs on its dedicated thread and events are used to notify the persister thread whenever
+	// transactions are added/removed from the mempool. The persister thread then updates the database accordingly.
+	persister  *MempoolPersister
+	emitEvents bool
+
+	// latestBlockView is used to check if a transaction is valid before being added to the mempool. The latestBlockView
+	// checks if the transaction has a valid signature and if the transaction's sender has enough funds to cover the fee.
+	// The latestBlockView should be updated whenever a new block is added to the blockchain via UpdateLatestBlock.
+	latestBlockView *UtxoView
+	// latestBlockNode is used to infer the latest block height. The latestBlockNode should be updated whenever a new
+	// block is added to the blockchain via UpdateLatestBlock.
+	latestBlockNode *BlockNode
+}
+
+func NewPosMempool(params *DeSoParams, globalParams *GlobalParamsEntry, latestBlockView *UtxoView,
+	latestBlockNode *BlockNode, dir string) *PosMempool {
+	return &PosMempool{
+		status:          PosMempoolStatusNotRunning,
+		params:          params,
+		globalParams:    globalParams,
+		dir:             dir,
+		emitEvents:      false,
+		latestBlockView: latestBlockView,
+		latestBlockNode: latestBlockNode,
+	}
+}
+
+func (dmp *PosMempool) Start() error {
+	if dmp.IsRunning() {
+		return nil
+	}
+
```

Comment:
This is a fun one. I think a mutex here will cause a deadlock. For instance, `GetPersistedTransactions` will deadlock. Rather than hacking these methods to support a no-lock variant, I think we can actually get away without locks in `Start` and `Stop`. The reason is that `status` is only changed at the very end of `Start` and `Stop`. This means that `Start` will only ever execute if node actually stoped or it was never run in the first place. Similarly, `Stop` will never execute, unless `Start` was run and finished executing. Because of this, I think `Start` and `Stop` are actually thread-safe and don't need locks. But see if you can think of any edge-case.